File: node-js/src/core/Puzzle2D.js
const Piece = require('./Piece');
const ExactCoverCounter = require('./ExactCoverCounter');

class Puzzle2D {
  constructor() {
    this.PIECES = 12;
    this.ROWS = 0;
    this.COLUMNS = 0;
    this.totalSolutions = 0;
    this.triedPieces = 0;
    this.row = 0;
    this.column = 0;
    this.piecesIndices = [];
    this.solution = new Array(this.PIECES);
    this.pieces = new Array(this.PIECES);
    this.names = '';
    this.grid = [];
    this.currPiece = null;
    this.totalFillInGrid = 0;
    this.availInGrid = 0;
    this._aborted = false;
    this.enumerateAllSolutions = false;
  }

  set(rows, columns) {
    Piece.totalFill = 0;
    this.piecesIndices = [];
    this.solution = new Array(this.PIECES);
    this.totalSolutions = 0;
    this.triedPieces = 0;
    this._aborted = false;

    this.ROWS = rows;
    this.COLUMNS = columns;

    this.grid = new Array(this.ROWS);
    for (let i = 0; i < this.ROWS; i++) {
      this.grid[i] = new Array(this.COLUMNS).fill(0);
    }

    const allPieces = [
      [[1], [1], [1], [1, 1]],
      [[1, 1], [1], [1, 1]],
      [[0, 1, 1], [1, 1], [0, 1]],
      [[0, 1], [1, 1, 1], [0, 1]],
      [[1, 1, 1, 1], [0, 0, 1]],
      [[0, 1], [1, 1], [1], [1]],
      [[0, 0, 1], [0, 1, 1], [1, 1]],
      [[1], [1, 1], [1, 1]],
      [[0, 0, 1], [1, 1, 1], [1]],
      [[0, 0, 1], [0, 0, 1], [1, 1, 1]],
      [[0, 0, 1], [1, 1, 1], [0, 0, 1]],
      [[1, 1, 1, 1, 1]]
    ];

    const rotations = [4, 4, 2, 1, 4, 4, 4, 4, 2, 4, 4, 2];
    const symmetric = [2, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 1];
    this.names = 'LUFXYNWPZVTI';

    if (this.ROWS === this.COLUMNS && this.ROWS === 8) {
      rotations[2] = 1;
    }

    if (this.ROWS === 0 && this.COLUMNS === 0) {
      throw new Error('not supported yet');
    } else {
      for (let i = 0; i < this.PIECES; i++) {
        this.piecesIndices.push(i);
        this.pieces[i] = new Piece(i, allPieces[i], rotations[i], symmetric[i], this.names[i]);
      }
      if (this.ROWS === this.COLUMNS && this.ROWS === 8) {
        this.grid[0][0] = -1;
        this.grid[7][0] = -1;
        this.grid[0][7] = -1;
        this.grid[7][7] = -1;
      }
      while (this.grid[0][this.column] === -1) {
        this.column++;
      }
    }

    this.totalFillInGrid = this.ROWS * this.COLUMNS;
    for (const row of this.grid) {
      for (const val of row) {
        if (val === -1) {
          this.totalFillInGrid--;
        }
      }
    }

    this.availInGrid = this.totalFillInGrid;
    console.log(`Found ${this.ROWS} rows, ${this.COLUMNS} cols, with total of cells ${this.availInGrid}`);
    this.showGrid();
  }

  showGrid() {
    console.log(`[msec] showGrid. Tried Pieces ${this.triedPieces} leftPieces ${this.piecesIndices.length}`);
    for (let i = 0; i < this.ROWS; i++) {
      let line = '';
      for (let j = 0; j < this.COLUMNS; j++) {
        if (this.grid[i][j] === -1) {
          line += '*  ';
        } else if (this.grid[i][j] === 0) {
          line += '-  ';
        } else {
          line += `${this.names[this.grid[i][j] - 1]} `;
        }
      }
      console.log(line);
    }
  }

  showPieces() {
    let line = '';
    for (let i = 0; i < this.PIECES - this.piecesIndices.length; i++) {
      line += `${this.names[this.solution[i]]} `;
    }
    console.log(line);
  }

  put() {
    const leftPieces = this.piecesIndices.length;

    if (leftPieces === 0) {
      this.totalSolutions++;
      console.log(`totalSolutions ${this.totalSolutions}`);
      if (this.totalSolutions === 1) {
        console.log('Found a solution');
        this.showGrid();
        this.showPieces();
      }
    }

    if (!this.enumerateAllSolutions && this.totalSolutions >= 1) return;

    const rowsSet = new Array(5);
    const columnsSet = new Array(5);

    for (let i = 0; i < leftPieces; i++) {
      const pieceIdx = this.piecesIndices[i];
      this.currPiece = this.pieces[pieceIdx];
      for (let r = this.currPiece.getAvailRotations(); r >= 1; r--) {
        if (this.canPut(rowsSet, columnsSet)) {
          this.piecesIndices.splice(i, 1);
          this.solution[this.PIECES - leftPieces] = pieceIdx;
          this.putCurrPiece(rowsSet, columnsSet);

          if (this.triedPieces % 50000 === 0) {
            this.showGrid();
            this.showPieces();

            if (this._aborted) {
              console.log(`Id ${this.ROWS}_${this.COLUMNS}. Signaled timed out! totalSolutions ${this.totalSolutions}`);
              return;
            }
          }

          this.put();
          this.removeLast(pieceIdx, rowsSet, columnsSet);
          this.piecesIndices.splice(i, 0, pieceIdx);
        }
        this.currPiece.rotate();
      }
    }
  }

  putCurrPiece(rowsSet, columnsSet) {
    this.currPiece.setPosition(this.row, this.column);

    const currIndex = this.currPiece.index + 1;
    for (let i = 0; i < this.currPiece.totalThisFill; i++) {
      this.grid[rowsSet[i]][columnsSet[i]] = currIndex;
    }

    this.goForward();
    this.availInGrid -= this.currPiece.totalThisFill;
  }

  goForward() {
    while (this.row < this.ROWS) {
      while (this.column < this.COLUMNS) {
        if (this.grid[this.row][this.column] === 0) return;
        this.column++;
      }
      this.column = 0;
      this.row++;
    }
  }

  removeLast(piece, rowsSet, columnsSet) {
    this.currPiece = this.pieces[piece];

    this.row = this.currPiece.getRow();
    this.column = this.currPiece.getColumn();

    for (let i = 0; i < this.currPiece.totalThisFill; i++) {
      this.grid[rowsSet[i]][columnsSet[i]] = 0;
    }

    this.availInGrid += this.currPiece.totalThisFill;
  }

  canPut(rowsSet, columnsSet) {
    try {
      let setSoFar = 0;
      const j = this.currPiece.getFirstSquarePos();
      const columnjj = this.column - j;
      for (let i = 0; i < this.currPiece.totalThisFill; i++) {
        const rowi = this.row + this.currPiece.getRowSet(setSoFar);
        const columnj = columnjj + this.currPiece.getColumnSet(setSoFar);
        if (this.grid[rowi][columnj] !== 0) {
          return false;
        } else {
          rowsSet[setSoFar] = rowi;
          columnsSet[setSoFar] = columnj;
          setSoFar++;
        }
      }
    } catch (_) {
      return false;
    }

    this.triedPieces++;
    return true;
  }

  buildExactCoverRows() {
    const cellColumns = new Array(this.ROWS);
    let cellColumn = 0;
    for (let row = 0; row < this.ROWS; row++) {
      cellColumns[row] = new Array(this.COLUMNS);
      for (let column = 0; column < this.COLUMNS; column++) {
        cellColumns[row][column] = this.grid[row][column] === -1 ? -1 : cellColumn++;
      }
    }

    const exactCoverRows = [];
    for (let pieceIndex = 0; pieceIndex < this.PIECES; pieceIndex++) {
      const piece = this.pieces[pieceIndex];
      for (let rotation = 0; rotation < piece.getAvailRotations(); rotation++) {
        const rows = piece.rowsSet[piece.currRotation];
        const columns = piece.columnsSet[piece.currRotation];
        const minRow = Math.min(...rows);
        const maxRow = Math.max(...rows);
        const minColumn = Math.min(...columns);
        const maxColumn = Math.max(...columns);

        for (let rowOffset = -minRow; rowOffset + maxRow < this.ROWS; rowOffset++) {
          for (
            let columnOffset = -minColumn;
            columnOffset + maxColumn < this.COLUMNS;
            columnOffset++
          ) {
            const exactCoverRow = [];
            let available = true;
            for (let cell = 0; cell < piece.totalThisFill; cell++) {
              const gridRow = rowOffset + rows[cell];
              const gridColumn = columnOffset + columns[cell];
              if (this.grid[gridRow][gridColumn] === -1) {
                available = false;
                break;
              }
              exactCoverRow.push(cellColumns[gridRow][gridColumn]);
            }
            if (available) {
              exactCoverRow.push(this.totalFillInGrid + pieceIndex);
              exactCoverRows.push(exactCoverRow);
            }
          }
        }
        piece.rotate();
      }
    }
    return exactCoverRows;
  }

  solve(signal, enumerateAllSolutions = false) {
    const start = Date.now();
    this.enumerateAllSolutions = enumerateAllSolutions;

    if (signal) {
      signal.addEventListener('abort', () => {
        this._aborted = true;
      });
    }

    if (this.totalFillInGrid !== Piece.totalFill) {
      console.error(`Invalid config, grid ${this.totalFillInGrid} pieces ${Piece.totalFill}`);
    } else if (enumerateAllSolutions) {
      const result = ExactCoverCounter.count(
        this.totalFillInGrid + this.PIECES,
        this.buildExactCoverRows()
      );
      this.totalSolutions = result.solutions;
      this.triedPieces = result.triedRows;
    } else {
      Piece.totalFill = 0;
      console.log(`Starting rows ${this.ROWS} cols ${this.COLUMNS}`);
      this.put();
    }

    const elapsedTime = Math.floor((Date.now() - start) / 1000);
    console.log(`tried ${this.triedPieces.toLocaleString()} pieces`);
    if (elapsedTime > 0) {
      console.log(`at ${(this.triedPieces / elapsedTime).toLocaleString()} pieces per sec`);
    }
    const reportedSolutions = enumerateAllSolutions
      ? this.totalSolutions
      : this.totalSolutions * (this.ROWS === this.COLUMNS ? 8 : 4);
    console.log(`number of solutions ${reportedSolutions}`);

    return this.totalSolutions;
  }
}

module.exports = Puzzle2D;

File: python/src/core/puzzle2d.py
from __future__ import annotations

import time
from typing import Optional

from .piece import Piece


class Puzzle2D:
    def __init__(self) -> None:
        self.PIECES = 12
        self.ROWS = 0
        self.COLUMNS = 0
        self.total_solutions = 0
        self.tried_pieces = 0
        self.row = 0
        self.column = 0
        self.pieces_indices: list[int] = []
        self.solution: list[int] = [0] * self.PIECES
        self.pieces: list[Optional[Piece]] = [None] * self.PIECES
        self.names = ""
        self.grid: list[list[int]] = []
        self.curr_piece: Optional[Piece] = None
        self.total_fill_in_grid = 0
        self.total_fill_in_pieces = 0
        self.avail_in_grid = 0
        self._aborted = False

    def set(self, rows: int, columns: int) -> None:
        # A rectangular board and its 90-degree rotation have the same tilings.
        # Search the taller orientation because the row-major backtracking order
        # is dramatically faster for narrow boards such as 20x3.
        if columns > rows:
            rows, columns = columns, rows

        self.pieces_indices = []
        self.solution = [0] * self.PIECES
        self.total_solutions = 0
        self.tried_pieces = 0
        self.total_fill_in_pieces = 0
        self._aborted = False

        self.ROWS = rows
        self.COLUMNS = columns

        self.grid = [[0] * self.COLUMNS for _ in range(self.ROWS)]

        all_pieces = [
            [[1], [1], [1], [1, 1]],
            [[1, 1], [1], [1, 1]],
            [[0, 1, 1], [1, 1], [0, 1]],
            [[0, 1], [1, 1, 1], [0, 1]],
            [[1, 1, 1, 1], [0, 0, 1]],
            [[0, 1], [1, 1], [1], [1]],
            [[0, 0, 1], [0, 1, 1], [1, 1]],
            [[1], [1, 1], [1, 1]],
            [[0, 0, 1], [1, 1, 1], [1]],
            [[0, 0, 1], [0, 0, 1], [1, 1, 1]],
            [[0, 0, 1], [1, 1, 1], [0, 0, 1]],
            [[1, 1, 1, 1, 1]],
        ]

        rotations = [4, 4, 2, 1, 4, 4, 4, 4, 2, 4, 4, 2]
        symmetric = [2, 1, 1, 1, 2, 2, 1, 2, 2, 1, 1, 1]
        self.names = "LUFXYNWPZVTI"

        if self.ROWS == self.COLUMNS and self.ROWS == 8:
            rotations[2] = 1

        if self.ROWS == 0 and self.COLUMNS == 0:
            raise ValueError("not supported yet")
        else:
            for i in range(self.PIECES):
                self.pieces_indices.append(i)
                self.pieces[i] = Piece(
                    i, all_pieces[i], rotations[i], symmetric[i], self.names[i]
                )
                self.total_fill_in_pieces += self.pieces[i].total_this_fill
            if self.ROWS == self.COLUMNS and self.ROWS == 8:
                self.grid[0][0] = -1
                self.grid[7][0] = -1
                self.grid[0][7] = -1
                self.grid[7][7] = -1
            while self.grid[0][self.column] == -1:
                self.column += 1

        self.total_fill_in_grid = self.ROWS * self.COLUMNS
        for row in self.grid:
            for val in row:
                if val == -1:
                    self.total_fill_in_grid -= 1

        self.avail_in_grid = self.total_fill_in_grid
        print(
            f"Found {self.ROWS} rows, {self.COLUMNS} cols, "
            f"with total of cells {self.avail_in_grid}"
        )
        self.show_grid()

    def show_grid(self) -> None:
        print(
            f"[msec] showGrid. Tried Pieces {self.tried_pieces} "
            f"leftPieces {len(self.pieces_indices)}"
        )
        for i in range(self.ROWS):
            line = ""
            for j in range(self.COLUMNS):
                if self.grid[i][j] == -1:
                    line += "*  "
                elif self.grid[i][j] == 0:
                    line += "-  "
                else:
                    line += f"{self.names[self.grid[i][j] - 1]} "
            print(line)

    def show_pieces(self) -> None:
        line = ""
        for i in range(self.PIECES - len(self.pieces_indices)):
            line += f"{self.names[self.solution[i]]} "
        print(line)

    def put(self) -> None:
        left_pieces = len(self.pieces_indices)

        if left_pieces == 0:
            self.total_solutions += 1
            print(f"totalSolutions {self.total_solutions}")
            if self.total_solutions == 1:
                print("Found a solution")
                self.show_grid()
                self.show_pieces()

        if self.total_solutions >= 1:
            return

        rows_set = [0] * 5
        columns_set = [0] * 5

        for i in range(left_pieces):
            piece_idx = self.pieces_indices[i]
            self.curr_piece = self.pieces[piece_idx]
            for r in range(self.curr_piece.get_avail_rotations(), 0, -1):
                if self.can_put(rows_set, columns_set):
                    self.pieces_indices.pop(i)
                    self.solution[self.PIECES - left_pieces] = piece_idx
                    self.put_curr_piece(rows_set, columns_set)

                    if self.tried_pieces % 50000 == 0:
                        self.show_grid()
                        self.show_pieces()

                        if self._aborted:
                            print(
                                f"Id {self.ROWS}_{self.COLUMNS}. "
                                f"Signaled timed out! totalSolutions {self.total_solutions}"
                            )
                            return

                    self.put()
                    self.remove_last(piece_idx, rows_set, columns_set)
                    self.pieces_indices.insert(i, piece_idx)
                self.curr_piece.rotate()

    def put_curr_piece(self, rows_set: list[int], columns_set: list[int]) -> None:
        assert self.curr_piece is not None
        self.curr_piece.set_position(self.row, self.column)

        curr_index = self.curr_piece.index + 1
        for i in range(self.curr_piece.total_this_fill):
            self.grid[rows_set[i]][columns_set[i]] = curr_index

        self.go_forward()
        self.avail_in_grid -= self.curr_piece.total_this_fill

    def go_forward(self) -> None:
        while self.row < self.ROWS:
            while self.column < self.COLUMNS:
                if self.grid[self.row][self.column] == 0:
                    return
                self.column += 1
            self.column = 0
            self.row += 1

    def remove_last(
        self, piece: int, rows_set: list[int], columns_set: list[int]
    ) -> None:
        assert self.pieces[piece] is not None
        self.curr_piece = self.pieces[piece]
        assert self.curr_piece is not None

        self.row = self.curr_piece.get_row()
        self.column = self.curr_piece.get_column()

        for i in range(self.curr_piece.total_this_fill):
            self.grid[rows_set[i]][columns_set[i]] = 0

        self.avail_in_grid += self.curr_piece.total_this_fill

    def can_put(self, rows_set: list[int], columns_set: list[int]) -> bool:
        assert self.curr_piece is not None
        try:
            set_so_far = 0
            j = self.curr_piece.get_first_square_pos()
            column_jj = self.column - j
            for i in range(self.curr_piece.total_this_fill):
                row_i = self.row + self.curr_piece.get_row_set(set_so_far)
                column_j = column_jj + self.curr_piece.get_column_set(set_so_far)
                if (
                    row_i < 0
                    or row_i >= self.ROWS
                    or column_j < 0
                    or column_j >= self.COLUMNS
                ):
                    return False
                if self.grid[row_i][column_j] != 0:
                    return False
                else:
                    rows_set[set_so_far] = row_i
                    columns_set[set_so_far] = column_j
                    set_so_far += 1
        except (IndexError, TypeError):
            return False

        self.tried_pieces += 1
        return True

    def solve(self, abort_event: Optional[object] = None) -> int:
        start = time.time()

        if abort_event is not None and hasattr(abort_event, "is_set"):
            # Simulate AbortSignal: if event is set, treat as aborted
            if abort_event.is_set():  # type: ignore[union-attr]
                self._aborted = True

        if self.total_fill_in_grid != self.total_fill_in_pieces:
            print(
                f"Invalid config, grid {self.total_fill_in_grid} "
                f"pieces {self.total_fill_in_pieces}"
            )
        else:
            print(f"Starting rows {self.ROWS} cols {self.COLUMNS}")
            self.put()

        elapsed_time = int(time.time() - start)
        print(f"tried {self.tried_pieces:,} pieces")
        if elapsed_time > 0:
            print(f"at {self.tried_pieces // elapsed_time:,} pieces per sec")
        print(
            f"number of solutions {self.get_solution_count()}"
        )

        return self.total_solutions

    def get_solution_count(self) -> int:
        """Return solutions including equivalent board symmetries."""
        symmetries = 8 if self.ROWS == self.COLUMNS else 4
        return self.total_solutions * symmetries
